File: Diagrammatize.Design.Pattern/code/cxx/AbstractFactory/A2/tablefactory/TableFactory.cpp
#include "TableFactory.hpp"
#include "TableLink.hpp"
#include "TableTray.hpp"
#include "TablePage.hpp"

Link* TableFactory::createLink(const std::string& caption, const std::string& url) {
    return new TableLink(caption, url);
}

Tray* TableFactory::createTray(const std::string& caption) {
    return new TableTray(caption);
}

Page* TableFactory::createPage(const std::string& title, const std::string& author) {
    return new TablePage(title, author);
}

static Factory* newTableFactory() {
    return new TableFactory();
}

//static FactoryRegister factory("TableFactory", &newTableFactory);
REGISTER_FACTORY(TableFactory, &newTableFactory);

File: Diagrammatize.Design.Patterns/code/python/AbstractFactory/A2/tablefactory/TableFactory.py
from factory import Factory
from .TableLink import TableLink
from .TableTray import TableTray
from .TablePage import TablePage

class TableFactory(Factory):
    def createLink(self, caption, url):
        return TableLink(caption, url)

    def createTray(self, caption):
        return TableTray(caption)

    def createPage(self, title, author):
        return TablePage(title, author)
